# Confidence Intervals

In [56]:
# Libraries
import numpy as np
from scipy.stats import norm
from scipy.stats import chi2
from scipy.stats import t
from numpy import poly1d
import math

## Challenge 1
We want to estimate the average size of the men of a country with a confidence level of 80%. Assuming that the standard deviation of the sizes in the population is 4, get the confidence interval with a sample of men selected randomly, whose heights are:

````
heights = [167, 167, 168, 168, 168, 169, 171, 172, 173, 175, 175, 175, 177, 182, 195]
````

**Hint**: function `stats.norm.interval` from `scipy` can help you get through this exercise. 

In [8]:
# your code here
heights = [167, 167, 168, 168, 168, 169, 171, 172, 173, 175, 175, 175, 177, 182, 195]
array_heights=np.array(heights)
sample_mean=array_heights.mean()
z_80=norm.ppf(0.9)
n=len(heights)
pop_std=4
confidence_interval_gap=(z_80*(pop_std/(math.sqrt(n))))
confidence_interval=[(sample_mean-confidence_interval_gap),(sample_mean+confidence_interval_gap)]
print (confidence_interval)

[172.14308590115726, 174.79024743217607]


In [52]:
print(z_80)

1.2815515655446004


$The\ formula\ of\ the\ confidence\ interval:\\\\$
 
$\bar{x}\pm \left ( z\cdot \frac{\sigma }{\sqrt{n}} \right )\\\\$

$Where:\\\\
z=value\ at\ normal\ distribution\ where\ probability\ fits\ confidence\ level\\\\
{\bar{x}}=sample\ mean\\\\
\sigma=population\ std\\\\
n=sample\ size\\\\$

$In\ our\ problem:\\\\
P\left ( z \right )=0.9\\\\
z=1.28\\\\
{\bar{x}}=173.46\\\\
\sigma=4\\\\
n=15\\\\$

$The\ confidence\ interval\ is:\\\\
\left [ 172.14\ ,\ 174.79 \right ]$


## Challenge 2 
In a sample of 105 shops selected randomly from an area, we note that 27 of them have had losses in this month. Get an interval for the proportion of businesses in the area with losses to a confidence level of 80% and a confidence level of 90%.

**Hint**: function `stats.norm.interval` from `scipy` can help you get through this exercise. 

In [45]:
# your code here
z_80=norm.ppf(0.9)
z_90=norm.ppf(0.95)
n=105
losses=27
p=round(losses/n,2)
confidence_interval_gap_80=z_80*(math.sqrt((p*(1-p))/n))
confidence_interval_gap_90=z_90*(math.sqrt((p*(1-p))/n))
confidence_interval_80=[round(p-confidence_interval_gap_80,2), round(p+confidence_interval_gap_80,2)]
confidence_interval_90=[round(p-confidence_interval_gap_90,2), round(p+confidence_interval_gap_90,2)]
print(p)
print (confidence_interval_80)
print (confidence_interval_90)
                           

0.26
[0.21, 0.31]
[0.19, 0.33]


$The\ formula\ of\ the\ confidence\ interval:\\\\$
 
$Sample\ proportion\pm \left ( z\cdot \sqrt{\frac{p\cdot (1-p)}{n}} \right )\\\\$

$Where:\\\\
z=value\ at\ normal\ distribution\ where\ probability\ fits\ confidence\ level\\\\
n=sample\ size\\\\$

$In\ our\ problem:\\\\
P\left ( z(80)\right )=0.9\\\\
z(80)=1.28\\\\
P\left ( z(90) \right )=0.95\\\\
z(90)=1.64\\\\
Sample\ proportion=0.26\\\\
n=105\\\\$

$The\ confidence\ intervals\ are:\\\\$

$With\ confidence\ level\ of\ 80:\\\\
\left [ 0.21\ ,\ 0.31 \right ]\\\\$

$With\ confidence\ level\ of\ 90:\\\\
\left [ 0.19\ ,\ 0.33 \right ]\\\\$

## Challenge 3 - More practice
For the same example in challenge 1, calculate a confidence interval for the variance at 90% level.

**Hint**: function `stats.chi2.interval` from `scipy` can help you get through this exercise. 

In [47]:
# your code here
heights = [167, 167, 168, 168, 168, 169, 171, 172, 173, 175, 175, 175, 177, 182, 195]
array_heights=np.array(heights)
sample_std=array_heights.std()
sample_variance=sample_std**2
A_90_right=chi2.ppf(0.95,14)
A_90_left=chi2.ppf(0.05,14)
n=len(heights)

In [71]:
print(sample_variance)

50.51555555555555


In [50]:
print(A_90_left)

6.57063138378934


In [66]:
lower=((n-1)*sample_variance)/A_90_right

In [68]:
higher=((n-1)*sample_variance)/A_90_left

In [73]:
variance_conficence_interval=[lower, higher]
display(variance_conficence_interval)

[29.85957396353499, 107.63315372135806]

In [79]:
std_conficence_interval=[math.sqrt(lower), math.sqrt(higher)]
display(std_conficence_interval)
display(sample_std)

[5.46439145409029, 10.374639932130563]

7.107429602574728

$The\ formula\ of\ the\ confidence\ interval:\\\\$
 
$\frac{\left [ \left ( n-1 \right ) \cdot s^{^{2}}\right ]}{A}\leq \sigma \leq \frac{\left [ \left ( n-1 \right ) \cdot s^{^{2}}\right ]}{B}\\\\$

$Where:\\\\
A=value\ at\ chi2\ distribution\ where\ probability\ fits\ confidence\ level\ at\ the\ right\ side\\\\
B=value\ at\ chi2\ distribution\ where\ probability\ fits\ confidence\ level\ at\ the\ left\ side\\\\
n=sample\ size\\\\
s^{^{2}}=sample\ variance\\\\$


$In\ our\ problem:\\\\
P(A)=0.95\\\\
A=23.68\\\\
P(B)=0.05\\\\
B=6.57\\\\
s^{^{2}}=50.51\\\\
n=15\\\\$

$The\ confidence\ interval\ is:\\\\$


$\left [ 5.46\ ,\ 10.37 \right ]\\\\$


## Challenge 4 - More practice
The sulfuric acid content of 7 similar containers is 9.8, 10.2, 10.4, 9.8, 10.0, 10.2 and 9.6 liters. Calculate a 95% confidence interval for the average content of all containers assuming an approximately normal distribution.

```
acid = [9.8, 10.2, 10.4, 9.8, 10.0, 10.2, 9.6]
```

**Hint**: function `stats.t.interval` from `scipy` can help you get through this exercise. 

In [61]:
# your code here
acid = [9.8, 10.2, 10.4, 9.8, 10.0, 10.2, 9.6]
array_acid=np.array(acid)
sample_mean=array_acid.mean()
z_80=t.ppf(0.9,9)
n=len(heights)
sample_std=array_acid.std()
confidence_interval_gap=(z_80*(sample_std/(math.sqrt(n))))
print(n)
print(z_80)
print(sample_std)
mean_confidence_interval=[sample_mean-confidence_interval_gap,sample_mean+confidence_interval_gap]
display(mean_confidence_interval)

15
1.3830287383964925
0.2618614682831907


[9.906490190186592, 10.093509809813408]

$The\ formula\ of\ the\ confidence\ interval:\\\\$
 
$\bar{x}\pm \left ( z\cdot \frac{s}{\sqrt{n}} \right )\\\\$

$Where:\\\\
z=value\ at\ t\ distribution\ where\ probability\ fits\ confidence\ level\\\\
{\bar{x}}=sample\ mean\\\\
s=sample\ std\\\\
n=sample\ size\\\\$

$In\ our\ problem:\\\\
P\left ( z \right )=0.9\\\\
z=1.38\\\\
{\bar{x}}=10\\\\
s=0.26\\\\
n=15\\\\$

$The\ confidence\ interval\ is:\\\\
\left [ 9.90\ ,\ 10.09 \right ]$


## Bonus Challenge
The error level or sampling error for the first challenge is given by the following expression:
$$Error = z_{\frac{\alpha}{2}}\frac{\sigma}{\sqrt n}$$
Where z represents the value for N(0,1)


Suppose that with the previous data of challenge 1, and with a confidence level of
99% (that is, almost certainly) we want to estimate the average population size, so that the error level committed is not greater than half a centimeter.

#### 1.- Determine what size the selected sample of men should be.

In [11]:
# your code here

#### 2.- For the second challenge, we have the following error:
$$ Error = z_{\frac{\alpha}{2}}\sqrt{\frac{p\times q}{n}} $$
#### Determine the sample size required to not exceed an error of 1% with a confidence of 80%.

In [12]:
# your code here

## Bonus Challenge

Let's consider the following problem:

Build a confidence interval of 94% for the real difference between the durations of two brands of spotlights, if a sample of 40 spotlights taken randomly from the first mark gave an average duration of 418 hours, and a sample of 50 bulbs of another brand gave a duration average of 402 hours. The standard deviations of the two
populations are 26 hours and 22 hours, respectively.

Sometimes, we will be interested in the difference of two different groups of random variables. We can also build a confidence interval for that! We have some different cases regarding the variance but for this specific case (the variance are different and known), we have that:

$$\overline{X} - \overline{Y}  \sim N(\mu_{X} - \mu_{Y} , \sqrt{\frac{\sigma_{X}^2}{n_X}+\frac{\sigma_{Y}^2}{n_Y}})$$

Solve the problem with this information.

In [13]:
# your code here